apiserver/common/networkingcommon/networkconfigapi.go

```diff
@@ -117,29 +129,229 @@ func (api *NetworkConfigAPI) getMachineForSettingNetworkConfig(machineTag string
+// longer observes locally.
+// The device itself is not marked for deletion now, but for later processing
+// to ensure it is not a parent of other observed devices.
+func (o *updateMachineLinkLayerOp) processExistingDeviceNotObserved(dev LinkLayerDevice) ([]txn.Op, error) {
+	addrs := o.DeviceAddresses(dev)
+
+	var ops []txn.Op
+	var removing int
+	for _, addr := range addrs {
+		// If the machine is the authority for this address,
+		// we can delete it; otherwise leave it alone.
+		if addr.Origin() == network.OriginMachine {
+			logger.Debugf("removing address %q from device %q", addr.Value(), dev.Name())
+			ops = append(ops, addr.RemoveOps()...)
+			removing++
+		}
+	}
+
+	// If the device is having all of its addresses removed and is not under
+	// the authority of the provider, add it as a candidate for removal.
+	// If the device has been relinquished by the provider, the instance-poller
+	// will have removed the provider ID - see the instance-poller API facade
+	// logic.
+	if removing == len(addrs) && dev.ProviderID() == "" {
+		o.removalCandidates = append(o.removalCandidates, dev)
+	}
+
+	return ops, nil
+}
+
+func (o *updateMachineLinkLayerOp) processExistingDeviceAddress(
+	dev LinkLayerDevice, addr LinkLayerAddress, incomingAddrs []state.LinkLayerDeviceAddress,
+) ([]txn.Op, error) {
+	addrValue := addr.Value()
+
+	// If one of the incoming addresses matches the existing one,
+	// update it.
+	for _, incomingAddr := range incomingAddrs {
+		if strings.HasPrefix(incomingAddr.CIDRAddress, addrValue) &&
+			!o.IsAddrProcessed(dev.MACAddress(), incomingAddr.CIDRAddress) {
+			o.MarkAddrProcessed(dev.MACAddress(), incomingAddr.CIDRAddress)
+
+			ops, err := addr.UpdateOps(incomingAddr)
+			return ops, errors.Trace(err)
+		}
+	}
+
+	// Otherwise if we are the authority, delete it.
+	if addr.Origin() == network.OriginMachine {
+		logger.Debugf("removing address %q from device %q", addrValue, addr.DeviceName())
+		return addr.RemoveOps(), nil
+	}
+
+	return nil, nil
+}
+
+// processExistingDeviceNewAddresses interrogates the list of incoming
+// addresses and adds any that were not processed as already existing.
+func (o *updateMachineLinkLayerOp) processExistingDeviceNewAddresses(
+	dev LinkLayerDevice, incomingAddrs []state.LinkLayerDeviceAddress,
+) ([]txn.Op, error) {
+	var ops []txn.Op
+	for _, addr := range incomingAddrs {
+		if !o.IsAddrProcessed(dev.MACAddress(), addr.CIDRAddress) {
+			addOps, err := dev.AddAddressOps(addr)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			ops = append(ops, addOps...)
+
+			o.MarkAddrProcessed(dev.MACAddress(), addr.CIDRAddress)
+		}
+	}
+	return ops, nil
+}
+
+// processNewDevices handles incoming devices that
+// did not match any we already have in state.
+func (o *updateMachineLinkLayerOp) processNewDevices() ([]txn.Op, error) {
+	var ops []txn.Op
+	for _, dev := range o.Incoming() {
+		if !o.IsDevProcessed(dev) {
```

Comment:
It might read a bit better to invert the logic here (also applies to the if !IsXProcessed calls in the other processX methods) to continue if something is processed and remove one level of indentation for the remainder of the code in the block.

Reply:
Done.